### PR TITLE
sequence protocol

### PR DESCRIPTION
**Pull request by palaviv**
Same as #765 but for `tuple` and `list`.

### CODE REVIEW

File: vm/src/obj/objsequence.rs [hunk 1]
```diff
@@ -137,7 +139,154 @@ impl<T: Clone> PySliceableSequence for Vec<T> {
     }
 }
 
-pub fn get_item(
+pub trait SequenceProtocol
+where
+    Self: Sized,
+{
+    fn get_elements(&self) -> Vec<PyObjectRef>;
```
Comment (adrian17) [inline vm/src/obj/objsequence.rs]: we definitely don't want to copy the entire vec just to eg check its size.

Comment (palaviv) [reply vm/src/obj/objsequence.rs]: Agree. I can't do a `Deref` inside the trait. Do you have a suggestion for something nicer then returning the `Refcell`?

Comment (coolreader18) [reply vm/src/obj/objsequence.rs]: You could return an `&[PyObjectRef]`, or if that wouldn't work an `impl Deref<Target = [PyObjectRef]>`.

File: vm/src/obj/objsequence.rs [hunk 2]
```diff
@@ -137,7 +139,154 @@ impl<T: Clone> PySliceableSequence for Vec<T> {
...
+    fn as_object(&self) -> &PyObjectRef;
+    fn into_object(self) -> PyObjectRef;
```
Comment (coolreader18) [inline vm/src/obj/objsequence.rs]: I think this could be split off into a separate trait, and we could have a generic `impl<T> ... for PyRef<T>`

**Comment by palaviv** [comment]
Should be implemented with inner like #785.

**Pull request closed without merging**